### DISCUSSION TITLE
Acquire line number at the point of the rule

### DISCUSSION THREAD

**Discussion by matejk** [foonathan/lexy#131] [Q&A] [answered]
Lexy is used to write a parser to read line-oriented syntax: each line can contain a different set of value/pair items. Lines are then processed like this:

```
    struct nc_file
    {
        static constexpr auto whitespace = dsl::ascii::space;
        static constexpr auto rule = [] {
            auto line = dsl::p<nc_line>;

            auto terminator = dsl::terminator(dsl::eof);
            return terminator.list(line);
        }();

        static constexpr auto value = lexy::as_list<std::vector<item_vector>>;
    };
```

I want to save the line number for each of the lines that are read. From the documentation of `input_location::line_nr()`, however I don't know how to use it in the rule or whether that is proper way to get the line number.

**Comment by foonathan** [comment, marked as answer]
The way it's supposed to work is that you use the rule [`dsl::position`](https://lexy.foonathan.net/reference/dsl/position/#position) which gives you an iterator into the input, then later call [`lexy::get_input_location()`](https://lexy.foonathan.net/reference/input_location/#get_input_location) to convert the iterators into line/column numbers.

However, in your particular case, it is probably easier if you use [`dsl::context_counter`](https://lexy.foonathan.net/reference/dsl/context_counter/#context_counter): create one in your `nr_file` rule, increment it after every line and use `.value()` to get the line number during parsing.

**Reply by natinusala** [reply]
I'm replying to this thread instead of making a new one because my question is related - if I have a rule like this: `LEXY_LIT("if") >> dsl::p<...>`

I can do this to get the position: `LEXY_LIT("if") >> dsl::position + dsl::p<...>`

But what if I want the position _before_ the if, do I need to use `peek`? `dsl::peek(LEXY_LIT("if")) >> dsl::position + LEXY_LIT("if") + dsl::p<...>`

Is this the intended way?

**Reply by foonathan** [reply]
Either do the `dsl::position` in the rule that has the choice (i.e.: `dsl::position + (dsl::p<if_stmt> | ... ))`) or use `dsl::capture(LEXY_LIT("if")) >> ...`, but note that it gives you a lexeme where you have to manually instruct the position with `.begin()`: https://lexy.foonathan.net/reference/dsl/capture/#capture
